### PR TITLE
Properly handle sealed classes with Kotlin 1.7 and JDK 17

### PR DESCRIPTION
**Pull request by stuebingerb**
Attempt to fix https://github.com/mockk/mockk/issues/832 by supporting sealed classes based on previous work done by @aSemy.

### CODE REVIEW

File: modules/mockk-agent-api/src/commonMain/kotlin/io/mockk/proxy/MockKInstantiatior.kt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 package io.mockk.proxy
 
 interface MockKInstantiatior {
-    fun <T> instance(cls: Class<T>): T
+    fun <T : Any> instance(cls: Class<T>): T
```
Comment (stuebingerb) [inline modules/mockk-agent-api/src/commonMain/kotlin/io/mockk/proxy/MockKInstantiatior.kt:4]: The `.kotlin` extension is only defined for `KClass<T : Any>` but from what I've seen it had been `T : Any` in all calling sites anyway.

Comment (stuebingerb) [reply modules/mockk-agent-api/src/commonMain/kotlin/io/mockk/proxy/MockKInstantiatior.kt:4]: Alright, seems I've missed something. Will revise.

**Comment by stuebingerb** [comment]
Currently it seems to _not_ solve our issues, though. We have the following class hierarchy (simplified):
```
sealed class User
abstract class KnownUser : User
class UserA : KnownUser
class UserB : KnownUser
```
and `getPermittedSubclasses` returns only the `KnownUser` class (which of course cannot be instantiated either).

**Comment by stuebingerb** [comment]
Currently looking into a potential open issue: although the mockk tests succeed, our own actual tests fail.

**Comment by aSemy** [comment]
Would you be able to add a test using the example that failed? (A sealed class with an abstract child?)

**Comment by stuebingerb** [comment]
> Would you be able to add a test using the example that failed? (A sealed class with an abstract child?)

Tried, but that test also succeeded without my modifications. I'm still missing something.

**Comment by aSemy** [comment]
Yes the stacktrace helps, thanks, although I can't figure out exactly what's wrong. At least I think it shows that we're on the right path! I don't see any reference to `ObjenesisInstantiator.kt`. The last MockK code called is in `io.mockk.proxy.jvm.transformation.SubclassInstrumentation`, so something in there (or earlier) isn't handling sealed subclasses properly. I suspect a similar fix will be needed.

**Comment by stuebingerb** [comment]
> Yes the stacktrace helps, thanks, although I can't figure out exactly what's wrong. At least I think it shows that we're on the right path! I don't see any reference to `ObjenesisInstantiator.kt`. The last MockK code called is in `io.mockk.proxy.jvm.transformation.SubclassInstrumentation`, so something in there (or earlier) isn't handling sealed subclasses properly. I suspect a similar fix will be needed.

My second commit here fixes the issue! I'm just wondering why and why I cannot come up with a dedicated test. Sorry for any confusion caused.

File: modules/mockk-agent/src/jvmMain/kotlin/io/mockk/proxy/jvm/ObjenesisInstantiator.kt [hunk 2]
```diff
@@ -24,6 +24,11 @@ class ObjenesisInstantiator(
         if (cls == Any::class.java) {
             @Suppress("UNCHECKED_CAST")
             return Any() as T
+        } else if (cls.isSealedSafe()) {
+            @Suppress("UNCHECKED_CAST")
+            return cls.getPermittedSubclassesSafe().firstNotNullOfOrNull { subCls ->
```
Comment (stuebingerb) [inline modules/mockk-agent/src/jvmMain/kotlin/io/mockk/proxy/jvm/ObjenesisInstantiator.kt:29]: @aSemy: Your PR did not have the `return` here and IIRC the mockk tests were also running successfully without. Was that an oversight or am I missing something?

Comment (aSemy) [reply modules/mockk-agent/src/jvmMain/kotlin/io/mockk/proxy/jvm/ObjenesisInstantiator.kt:29]: Good spot! Yes that's a mistake, I intended to use a `return`. However, it does show that Objenesis can probably handle sealed classes and we don't need to have custom code.

```kotlin
    private val instantiators = Collections.synchronizedMap(WeakHashMap<Class<*>, ObjectInstantiator<*>>())
    override fun <T> instance(cls: Class<T>): T {
        if (cls == Any::class.java) {
            @Suppress("UNCHECKED_CAST")
            return Any() as T
        } else if (!Modifier.isFinal(cls.modifiers) && !cls.isSealedSafe()) {
            try {
                val instance = instantiateViaProxy(cls)
                if (instance != null) {
                    return instance
                }
            } catch (ex: Exception) {
                log.trace(
                    ex, "Failed to instantiate via proxy " + cls + ". " +
                            "Doing objenesis instantiation"
                )
            }
        }
        return instanceViaObjenesis(cls)
    }
```

Comment (stuebingerb) [reply modules/mockk-agent/src/jvmMain/kotlin/io/mockk/proxy/jvm/ObjenesisInstantiator.kt:29]: Thought so as well - and the mockk tests are indeed running successfully. But in our code that leads to the original `InstantiationError`:

```
java.lang.InstantiationError: org.example.SealedClass
        at jdk.internal.reflect.GeneratedSerializationConstructorAccessor7.newInstance(Unknown Source)
        at java.base/java.lang.reflect.Constructor.newInstanceWithCaller(Constructor.java:499)
        at java.base/java.lang.reflect.Constructor.newInstance(Constructor.java:483)
        at org.objenesis.instantiator.sun.SunReflectionFactoryInstantiator.newInstance(SunReflectionFactoryInstantiator.java:48)
        at io.mockk.proxy.jvm.ObjenesisInstantiator.instanceViaObjenesis(ObjenesisInstantiator.kt:83)
        at io.mockk.proxy.jvm.ObjenesisInstantiator.instance(ObjenesisInstantiator.kt:42)
        at io.mockk.impl.instantiation.JvmInstantiator$instantiate$2.invoke(JvmInstantiator.kt:16)
```
I don't yet have a theory _why_ but I can see that our code is coming from the subclass creation.

But it looks like actually the changes in `ProxyMaker` are enough to get our tests running. I am still struggling to create a corresponding test case in mockk, though. Will update this PR later.

**Comment by stuebingerb** [comment]
Sorry for commit spam. The current implementation here now seems to fix both, mockk's tests and our own.

**Comment by stuebingerb** [comment]
So... I tried to add a new test case based on our code and I'm claiming that the CI tests are currently broken. Have a look at https://github.com/stuebingerb/mockk/pull/10 where the `SealedClassTest` and `SealedInterfaceTest` are both running successfully without any adjustments to the implementation.

@aSemy Any idea what might be wrong there?

**Comment by aSemy** [comment]
That is indeed weird! Good catch. I think the new Gradle GHA is re-using the cache for different JDK versions. I see this message in the logs: `Task :modules:mockk:test SKIPPED`

**Comment by stuebingerb** [comment]
@Raibaz @aSemy What is needed to get this merged?

**Comment by aSemy** [comment]
I thought that because there was an issue with the Gradle tests not failing https://github.com/mockk/mockk/pull/916#issuecomment-1240443435 that it wasn't clear if this PR would work as expected.

I don't have any suggestions on how to move forward...

**Comment by stuebingerb** [comment]
I guess someone will have to fix the CI. But I currently cannot.

**Comment by aSemy** [comment]
Let's do it then 👍

**Review by aSemy — APPROVED** [review]

**Comment by Raibaz** [comment]
Ok makes sense, sorry I lost track of what was going on here :)

**Comment by DHosseiny** [comment]
This change breaks mocking a sealed class with kotlin 1.6

**Pull request merged**